File: HOT100/10.回溯/78.子集.cpp
/*
 * @lc app=leetcode.cn id=78 lang=cpp
 *
 * [78] 子集
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        auto dfs = [&](auto&& self, int sta, vector<vector<int>>& res, vector<int>& pack) -> void
        {
            res.emplace_back(pack);
            for (int i = sta; i < n; i ++ )
            {
                pack.push_back(nums[i]);
                self(self, i + 1, res, pack);
                pack.pop_back();
            }
        };
        vector<vector<int>> res;
        vector<int> pack;
        dfs(dfs, 0, res, pack);
        return res;
    }
};
// @lc code=end

File: HOT100/10.2/78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python3
#
# [78] 子集
#
from mytools import *
# @lc code=start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        pack = []
        n = len(nums)
        def dfs(sta):
            res.append(pack[:])
            for i in range(sta, n):
                pack.append(nums[i])
                dfs(i + 1)
                pack.pop()
        dfs(0)
        return res
# @lc code=end

